### PR TITLE
Kimberley Mackenzie - Carets

### PR DESCRIPTION
**Pull request by kimberleyamackenzie**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? |   As I was designing, I thought about the idea of reserving a finite number of rooms from a finite but specific number of dates, eg a calendar.  For the idea of a hotel, creating a calendar with a year of date objects and attaching each of those dates to the 20 rooms of the hotel and their IDs and statuses on those dates, made the most sense to me.  As I worked, the other option of a more 'order-based' design became clear (where a date essentially only existed in the program once a room was reserved for that date).   Ultimately, I went with the calendar because it matched the real world concept of how a hotel would make reservations in my mind - drawing information from existing states, rather than generating states when given information.      |
| Describe a concept that you gained more clarity on as you worked on this assignment. |  I definitely got a deeper understanding of design in general and how the designing process & decisions can really define what/how you work for the remainder of that project.  I also feel like I gained a better understanding of why class variables aren't ideal.  I would love to hear about ways I could refactor my code- in particular the loop to iterate through the calendar takes up a chunk of lines and is used repeatedly throughout my classes, but the innermost loop changes depending on the task at hand, so using super/inheritance the way we have discussed them wouldn't work.  Is there a way I could have made this less repetitive? |
| Describe a _nominal test_ that you wrote for this assignment. |  Testing that a typical new block was created for a standard number of rooms for a given date range.   |
| Describe an _edge case test_ that you wrote for this assignment. |  In testing blocks, I tested a case where a block ID was entered that wasn't unique, which raised an error.  I also tested in reservations for overlapping reservations for the same room, which also raised an error.  |
| How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I really committed to this and found it a really helpful process to better understand the functionality of my methods and how they fit into the project as a whole.  |

### CODE REVIEW

File: lib/availability.rb [hunk 1]
```diff
@@ -0,0 +1,130 @@
+require_relative 'rooms'
+require 'date'
+
+class Availability
+  @@calendar = []
```
Comment (CheezItMan) [inline lib/availability.rb]: This class is an interesting idea, but it means you're keeping track of days, that don't have any reservations.  It's a bit memory inefficient.  

Also this means you can only have a program running one Hotel/Availability at a time.

File: lib/availability.rb [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+  # Creates a calendar array that includes a Date object key (for each day from generation until one year later) with a value of an array of each room and its status on that day
+  def self.create_calendar
+    current_date = Date.today
+    last_day = current_date + 366
+
+    roominfo = []
+    hotel = Hotel.new
+    hotel.rooms.each do |room|
```
Comment (CheezItMan) [inline lib/availability.rb]: This mixes details of the Hotel class implementation with the Calendar.  You're **tightly couplings** these classes, something Sandy Meitz would tell you to avoid.  

Instead the hotel object, when it's created should generate it's rooms.

File: lib/availability.rb [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+      id = room.id
+      status = room.status
+      roominfo << {id => status}
+    end
+
+    until current_date == last_day
+      roominfo = Marshal.load(Marshal.dump(roominfo))
```
Comment (CheezItMan) [inline lib/availability.rb]: Neat research using `Marshal` to save and load data from files.

File: lib/availability.rb [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+    wanteddate = checkin_date
+
+    until wanteddate == checkout_date
+      self.calendar.each do |days|
+        days.each do |date, roominfo|
+          if wanteddate == date
+            roominfo.each do |rooms|
+              rooms.each do |id, status|
+                if status == :available
```
Comment (CheezItMan) [inline lib/availability.rb]: What if a room is available for 1 day in the range?  Could that room end up getting added to the list?

File: specs/availability_spec.rb [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    it "shows each day for year including and following the day the availability is called" do
+      date = Date.today
+      nextyear = date + 365
+      @all_availability[0].keys[0].must_equal date
+      @all_availability[365].keys[0].must_equal nextyear
+    end
+  end
+
+  describe "all available rooms method" do
+    it "returns an array of room ids" do
```
Comment (CheezItMan) [inline specs/availability_spec.rb:34]: There's no test here to ensure that if a room has been reserved in the range, the room doesn't appear.

File: lib/blocks.rb [hunk 6]
```diff
@@ -0,0 +1,56 @@
+require_relative 'reservations'
+require 'date'
+
+class Block
+  attr_reader :total_cost, :block_id, :blocked_rooms, :checkin_date, :checkout_date
+  @@every_block = []
+
+  # Create a new block by iterating through the calendar for the given date range and taking the first x rooms with a status of available and changing their status to :blocked
+  def initialize(room_count, startyear, startmonth, startday, endyear, endmonth, endday, block_id)
```
Comment (CheezItMan) [inline lib/blocks.rb:9]: This method is really large, it needs to be broken up.

File: lib/blocks.rb [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    if room_count > 5 || room_count.class != Integer || room_count < 2
+      raise ArgumentError.new("Rooms must be blocked in quantities between 2 and 5.")
+    end
+
+    Block.all_blocks.each do |block|
+      if block.block_id == block_id
```
Comment (CheezItMan) [inline lib/blocks.rb:15]: Why would the user need to create an ID that is unique.  Instead have the initialize assign one.

File: lib/blocks.rb [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    end
+
+    @block_id = block_id
+    @blocked_rooms = Availability.all_available_rooms(startyear, startmonth, startday, endyear, endmonth, endday)[0...room_count]
+
+    @checkin_date = Date.new(startyear,startmonth,startday)
+    @checkout_date = Date.new(endyear,endmonth,endday)
+    wanteddate = checkin_date
+
+    until wanteddate == checkout_date
```
Comment (CheezItMan) [inline lib/blocks.rb:27]: This type of loop is running many times.  Some kind of helper method would be useful.

File: lib/hotel.rb [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+    id = 1
+    20.times do
+      @rooms << Room.new(id)
+      id += 1
+    end
+  end
+
+  def list
+    return @rooms
```
Comment (CheezItMan) [inline lib/hotel.rb:24]: Why have a `rooms` accessor if you use this method?

File: lib/reservations.rb [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    end
+
+    total_stay = (checkout_date - checkin_date).to_i
+    @total_cost = total_stay * 200
+  end
+
+  # Reserve a blocked room from a specific block
+  def self.reserve_from_block(id)
+    chosen_block = ""
+    Block.all_blocks.each do |block|
```
Comment (CheezItMan) [inline lib/reservations.rb]: This is another example of tightly coupled classes.

File: specs/block_spec.rb [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+  end
+
+  describe "all blocks method" do
+    it "returns an array of all blocks" do
+      Block.new(4, 2017, 9, 28, 2017, 9, 29, "Briggs")
+      Block.all_blocks.must_be_instance_of Array
+    end
+  end
+
+  def teardown
```
Comment (CheezItMan) [inline specs/block_spec.rb:58]: Good work since you use class variables to reset the calendar between test runs.

**Comment by CheezItMan** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	|  You didn't do many commits here.  The messages are good however. |
Answer comprehension questions	| Check, I'm glad this was a good learning experience.   |
**Design** |
Each class is responsible for a single piece of the program | This is true somewhat, but the Availability class is both a calendar and a reservation checker.  This is a class with many responsibilities. |
Classes are loosely coupled | There's a lot of entanglement between classes here. See the notes I left in your code. Changing one class would require adjusting on another. Instead you should have each class interact with others through public methods.   |
**Wave 1** |
List rooms | Yes through the `rooms` accessor.   |
Reserve a room for a given date range | By creating a new `Reservation` instance. |
List reservations for a given date | Through `Availability.all_reservations` |
Calculate reservation price | Check |
Invalid date range produces an error | Yes through `Reservation` class' `initialize`.   |
**Wave 2** |
View available rooms for a given date range | Check in `Availability` class. |
Reserving a room that is not available produces an error | Check |
**Wave 3** |
Create a block of rooms | Check |
Check if a block has rooms | Check |
Reserve a room from a block | Check |
Test coverage | 99%!  Nice! |
**Additional Feedback** |  You got all the main functionality, but your design mixes functionality, tightly couples classes and thus would be very hard to adjust for changing requirements.  You're also using class variables which means your solution wouldn't work for example with multiple hotels.  Take a look at the comments I left in your code.  |

File: specs/reservations_spec.rb [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+    end
+
+    it "allows you to book a room beginning on the ending date of the previous resveration" do
+      reservation = Reservation.new(5, 2017, 10, 15, 2017, 10, 16)
+      reservation.must_be_instance_of Reservation
+      next_reservation = Reservation.new(5, 2017, 10, 16, 2017, 10, 17)
+      next_reservation.must_be_instance_of Reservation
+    end
+
+    it "changes the status of a room to booked and does not allow another reservation to overlap" do
```
Comment (CheezItMan) [inline specs/reservations_spec.rb:21]: What about testing for invalid dates?

File: specs/reservations_spec.rb [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+      reservation = Reservation.new(1, 2017, 9, 28, 2017, 9, 29)
+      reservation.total_cost.must_be_instance_of Integer
+      reservation.total_cost.must_equal 200
+      another_reservation = Reservation.new(2, 2017, 9, 28, 2017, 10, 1)
+      another_reservation.total_cost.must_be_instance_of Integer
+      another_reservation.total_cost.must_equal 600
+    end
+  end
+
+  describe "reserves from block" do
```
Comment (CheezItMan) [inline specs/reservations_spec.rb:36]: You have nominal tests here, but no edge cases.

File: specs/hotel_spec.rb [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    it "gives all rooms a default rate of 200" do
+      @hotel.rooms[0].rate.must_equal 200
+      @hotel.rooms[19].rate.must_equal 200
+      @hotel.rooms[9].rate.must_equal 200
+    end
+  end
+
+  describe "list method" do
+    it "generates a list of all rooms in the hotel" do
+      @hotel.list.must_be_instance_of Array
```
Comment (CheezItMan) [inline specs/hotel_spec.rb:44]: You should probably also check to ensure that all the elements are rooms.

File: specs/block_spec.rb [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+      proc { Block.new(3, 2017, 9, 28, 2017, 9, 29, "Brennan")}.must_raise ArgumentError
+    end
+
+    it "returns an array of rooms that have been blocked" do
+      block = Block.new(4, 2017, 9, 28, 2017, 9, 29, "Cooper")
+      block.blocked_rooms.must_be_instance_of Array
+      block.blocked_rooms.count.must_equal 4
+    end
+
+    it "has a single check-in and check-out date that applies to all rooms in the block" do
```
Comment (CheezItMan) [inline specs/block_spec.rb:35]: You need to check edge cases, like invalid days, repeated reservations for the same room, reserving all the rooms etc.

**Review by CheezItMan — COMMENTED** [review]
Some comments on the project.
